File: android/src/main/java/com/margelo/nitro/rive/HTTPDataLoader.kt
package com.margelo.nitro.rive

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

object HTTPDataLoader : DataLoader {
  override suspend fun load(source: DataSource): ByteArray {
    val httpSource = source as? DataSource.Http
      ?: throw DataLoaderException.InvalidSource
    return downloadBytes(httpSource.url)
  }

  suspend fun downloadBytes(url: String): ByteArray = withContext(Dispatchers.IO) {
    val urlObj = try {
      URL(url)
    } catch (e: MalformedURLException) {
      throw DataLoaderException.InvalidURL(url)
    }
    val connection = urlObj.openConnection() as HttpURLConnection

    try {
      connection.requestMethod = "GET"
      val statusCode = connection.responseCode

      if (statusCode !in 200..299) {
        throw DataLoaderException.HttpError(statusCode, url)
      }

      connection.inputStream.use { it.readBytes() }
    } finally {
      connection.disconnect()
    }
  }
}

File: ios/HTTPDataLoader.swift
import Foundation

final class HTTPDataLoader: DataLoader {
  static let shared = HTTPDataLoader()

  func load(from source: DataSource) async throws -> Data {
    guard case .http(let url) = source else {
      throw DataLoaderError.invalidSource
    }
    return try await downloadData(from: url)
  }

  func downloadData(from url: URL) async throws -> Data {
    let (data, response) = try await URLSession.shared.data(from: url)

    if let httpResponse = response as? HTTPURLResponse,
      !(200...299).contains(httpResponse.statusCode)
    {
      throw DataLoaderError.httpError(statusCode: httpResponse.statusCode, url: url)
    }

    return data
  }

  func downloadData(from urlString: String) async throws -> Data {
    guard let url = URL(string: urlString) else {
      throw DataLoaderError.invalidURL(urlString)
    }
    return try await downloadData(from: url)
  }
}
